### ISSUE TITLE
DivideByZero exception has conflicting attributes

### ISSUE THREAD

**Issue by coke** [Raku/problem-solving#434]
While roast only seems to check the type of the exception, current rakudo has 3 attributes:

* numerator
* using
* details

Checking where the exception is thrown by rakudo, it seems to either be a numerator/using pair, the single attribute details, or nothing. We should be consistent in usage and probably select `numerator` and `using`, remove `details`, and ensure that wherever we throw the exception, we include the information.

Because there is no testing of these attributes in roast, we could do this as an implementation detail in existing rakudo, but I would recommend implementing it as a 6.e feature and then testing the attributes in roast for 6.e

**Comment by raiph** [comment]
I'm on vacation, and can't yet give this full attention, but [Raku's DBZ, successes, failures, DBZ exceptions, ...](https://gist.github.com/raiph/c5105d170cddcd7323c5bbdce0893e57).

**Comment by librasteve** [comment]
[please can you point to the source and I would be happy to try and grok the intent if I can]

for a long time IEEE P754 has given us valid values for the results of DBZ
these were provided to allow pipelined vector FPU ops to succeed (even if an interim result was Inf or NaN)
raku, uniquely [afaik], seeks to bridge this idea with Rats which can have a 0 denominator unless collapsed eg into a Str

following raiph's pointer to the docs on ZBRs, I see that this is OK:

```raku
say  <1/0>.Num;   # OUTPUT: «Inf␤» 
say <-1/0>.Num;   # OUTPUT: «-Inf␤» 
say  <0/0>.Num;   # OUTPUT: «NaN␤» 
say Inf.Rat.nude; # OUTPUT: «(1 0)␤» 
```

yet this is an error:

```raku
say <1/0>;
# Attempt to divide by zero when coercing Rational to Str in ...
```

my guess on the error attributes is:

- numerator is the numerator
- using is the denominator
- details reveals the mechanism, such as 
  - was the divide attempt in the CPU or the FPU
  - am I collapsing a Rat or reading across an extant FPU ±Inf
  - was the operation `div`, `/`, `.Str` (yes stringifying a Rat will try to print eg. a decimal)

**Comment by coke** [comment]
You can find all the instances (though you'll have to look at the source because many of these are part of a multi-line invocation) with

```
$ cd rakudo
$ ack 'X::Numeric::DivideByZero' src
```

`numerator` is the numerator (only ever specified with `using`). `using` is typically the operator that was called (e.g. `div`). _There's no need to specify the denominator... because it's `0`._

`details` is just plain text describing the situation, e.g.:

* `when calling .$what on Rational`
* `when coercing Rational to Str`

There is nothing in the current usage of rakudo or roast that requires `details` over the `numerator`/`using` combo. Both of the examples above are easily translatable (esp. since `using` is just free text)

There is **no** tie in to the CPU/FPU operations here.

**Comment by lizmat** [comment]
`s/ack/rak/`  ??  :-)

**Comment by librasteve** [comment]
@coke thanks for clarifying my guess - looks like the info I guessed at is split between the `using` and `details` field (as you say no need for denominator - red face)

Based on this, I see no issue in keeping the fields as they are since both the `using` (ie the operator) and the `details` (ie when doing what) seem useful for debugging. The ZBRs are a quite thorny and unique(?) to raku so for me at least finding any extra hint is very valuable to work out what is happening.

On the FPU, apologies if I did not explain myself carefully, but the situation I wanted to clarify (at least to my own satisfaction) is that where this is a thing:

```
say  <1/0>.Num;   # OUTPUT: «Inf␤» 
say  <1/0>.Num.WHAT;   # OUTPUT: «(Num)␤» 
```

And this (although you can't make a string from it):

```
say  <1/0>.Num.Rat.WHAT;   # OUTPUT: «(Rat)␤» 
```

Then, from a raku coder POV a valid mental model is that this is a divide by zero "non-failure" aka a ZBR that is made by creating a Inf (ie. by coercing a ZBR to an IEEE P754 Inf in the FPU since Num s are doubles) and then coercing it back to a Rat which as a construct of Ints lives in the CPU.

And I can make the identical thing via an FPU overflow like this:

```
say 1.7976931348623157e309;  # OUTPUT: «Inf␤» 
say 1.7976931348623157e309 === <1/0>.Num; # OUTPUT: «True␤» 
```

My point is that when you have done a bunch of P754 coding then it is most natural to think of Inf (aka the result of a DBZ op) as a creature of the FPU. This is the design of raku Num and the cool thing is the "interop" between Num and Rat - but also a source of potential complexity around DBZ behaviours.

So, obviously this would work too:

```
say (1e0/0e0).Rat; # OUTPUT: «Attempt to divide 1 by zero using /␤»...
```

.oO Realises that raku still supports DBZ exceptions in the FPU. Probably have to monkey with the compiler directives to put FPU in the "don't raise exception" mode atm. Would be nice to have that control on the same footing as eg `my $*RAT-OVERFLOW` one day.

It took me a while to realise that this Failure is from an FPU exception so maybe the `details` field could mention that? At least if not for now, then maybe this is the kind of nuanced message that could be put in `details` in future.

**Comment by librasteve** [comment]
there is a precursor to this problem-solving that was mooted for v6 
https://github.com/Raku/old-design-docs/blob/2c8786a52edb1970df7bb600968771aa03ed37f4/v6d.pod#use-ieee-754-2008-semantics-for-numnum-infix-infix-and-infix

**Comment by librasteve** [comment]
FWIW I have now implemented a "poor man's" version of 1/0 => Inf in App::Crag, since I think that this is better than a Failure for a CLI calculator

This raises my hope that this change will get attention, since to do this properly really needs the sign of numerator and denominator to select +/- Inf.

**Comment by raiph** [comment]
@librasteve 

> This raises my hope that this change will get attention, since to do this properly really needs the sign of numerator and denominator to select +/- Inf.

When you say "this change" are you talking about the change proposed in this issue (which supposes there are "conflicting" attributes in a DBZ exception, and that one of the three is redundant, and, if so, then it makes sense to eliminate the redundancy) or something to do with the "precursor to this issue" (which looks to me like an unrelated issue, specifically dividing a float by another float), or yet another issue?

**Comment by librasteve** [comment]
@raiph good question

So, there are two things in play:
 1. @coke asks why three fields (`numerator`, `using`, `details`)
 2. @Zoffix asks for more better DBZ features

I am revisiting this Issue, since I want [2]. Unless you say otherwise, I will open a new problem solving issue for [2]. Happy to keep this focused on [1].

I have spelunked a little and found this:

``` src/core.c/Num.rakumod
...
380 multi sub infix:</>(Num:D $a, Num:D $b) {
381     nqp::iseq_n($b,0e0) ?? $a.divzero !!  nqp::p6box_n(nqp::div_n($a,$b))
382 }   
...
```

I gather from this that `nqp:iseq` is testing if denominator == 0 and then issuing the DBZ.

If we only care about current features for [1], then I guess that `details` carrries no additional information.

---

That said, to implement [2] properly, you will need to consider the sign of the denominator also. Perhaps other info such as the exception source (CPU or FPU) or the computed result (eg NaN) would be useful too.

That's because the IEEE P754 standard handles signs like this:

`1e0/0e0  ==> Inf`
`-1e0/0e0  ==> -Inf`
`1e0/-0e0  ==> -Inf`
`-1e0/-0e0  ==> Inf`

`0e0/0e0  ==> NaN`
`Inf/Inf ==> NaN`

---

So yes let's move to a PR for [1] for this issue - this comment just a note to the implementers that a low level design that is open to these improvements would be cool.

**Comment by raiph** [comment]
**TL;DR** There's clearly redundancy. Why? I think Zoffix deliberately introduced the redundancy with the long term intent that Rakoons would eventually drop `.using` and replace it with `.details`. The motivation was a combination of a thorny scenario (when an error occurs in internal processing) and a minor detail (use of `.using` always inserts `'using '`).

----

[Zoffix's commit adding `.details`](https://github.com/rakudo/rakudo/commit/e22508d41a283df9987c3839b7fa6dfbf645597c):

> We don't always end up dividing by zero in user's code, so the "using..." bit isn't always appropriate.
>
> Add `$.details` attribute that will allow us to specify details on _why_ the divide by zero happened.

In the rest of this comment I dig into what I think they meant based on the four methods they added `.details` to.

----

The general issue is one of two distinct scenarios in which error messages arise:

* Error messages due to user directed operations, and their results, when the user _directly_ initiated that operation via their code, and can be expected to know about it, or at least be OK thinking about that operation and its results.

* Error messages due to _internal_ operations, and their results, that a user did _not_ directly initiate, and may well know _nothing_ about.

Obviously this distinction applies much more broadly than just DBZ exceptions. But DBZs can be a particularly sharp example, as I get into next.

----

There are four methods in Rational.rakumod (`.floor`, `.ceiling`, `.Int`, and `.round`) that make use of `.details`. Importantly, _other_ methods in Rational call one or more of the those four methods. In particular, the public methods `.Str` and `.base`. (Two private methods also do: `!STRINGIFY` and `!base`.)

As I understand things, in the case of both `.Str` and `.base`, the numerator may be mostly or entirely irrelevant, and the operation that leads to a DBZ exception may be entirely unknown by the user. For example, I imagine user code of the likes of `foo.base(8)` or `say bar;` might have ended with an `Attempt to divide 1427 by zero using .floor` error message in the old scheme (and `Attempt to divide by zero when calling .floor` in the new one).

I agree that the difference between the messages is so slight as to be negligible. But hopefully you can see that, of the two attributes, it's easy to see that `.details`, which does not inject any text, is more flexible than `.using`, which injects the string `'using '` in the middle of the error message, forcing the very concept of 'using', and naming that thing being used, when it may be something that a user did not use or know about.

----

FWIW I can see calls of the four methods that use `.details` coming from routines and operators in other packages, eg from operators like `%`. But for this comment I've stuck with stuff that happens entirely within the Rational package to keep things simple.

----

To close...

While you're thinking there's no good reason to have the `.using` vs `.details` redundancy, and that it's simplest to just keep `.using` and remove `.details`, I thought it was worth at least understanding _why_ Zoffix introduced the `.details` attribute.

I suspect they were thinking that it was a worthwhile win to eventually do the reverse -- to drop the _almost_ freeform `.using` attribute (that unfortunately unavoidably inserts the word "using " before the `.using` attribute's contents) and replace it with the _entirely_ freeform `.details` attribute.

Yes, that means prepending the string 'using ' to the strings passed as the values of `.using`, but it looks to me like it's pretty clearly what Zoffix was thinking.
